hasher-matcher-actioner/webapp/src/settings/ActionRuleSettingsTab.jsx

```diff
@@ -4,12 +4,93 @@
 
 /* eslint-disable react/prop-types */
 
-import React from 'react';
+import React, {useState} from 'react';
+import Button from 'react-bootstrap/Button';
 import Col from 'react-bootstrap/Col';
 import Container from 'react-bootstrap/Container';
+import Form from 'react-bootstrap/Form';
 import Row from 'react-bootstrap/Row';
+import Table from 'react-bootstrap/Table';
+
+const actionRules = [
+  {
+    name: 'Bank ID 303636684709969 Except Sailboat',
+    must_have_labels:
+      'BankIDClassificationLabel(303636684709969), ClassificationLabel(true_positive)',
+    must_not_have_labels:
+      'BankedContentIDClassificationLabel(3364504410306721)',
+    action_id: '1',
+  },
+  {
+    name: 'Bank ID 303636684709969, Sailboat',
+    must_have_labels:
+      'BankIDClassificationLabel(303636684709969), ClassificationLabel(true_positive), BankedContentIDClassificationLabel(3364504410306721)',
+    must_not_have_labels: '',
+    action_id: '2',
+  },
+];
+
+const actions = [
+  {
+    name: 'EnqueueMiniCastleForReview',
+    id: '1',
+  },
+  {
+    name: 'EnqueueSailboatForReview',
+    id: '2',
+  },
+];
+
+function ActionRuleFormColumns(props) {
+  const {nameRef, mustHaveLabelsRef, mustNotHaveLabelsRef, actionRef} = props;
+
+  const actionOptions = actions.map(action => (
+    <option key={action.id} value={action.id}>
+      {action.name}
+    </option>
+  ));
+
+  return (
+    <>
+      <td>
+        <Form.Control type="text" required ref={nameRef} />
+        <Form.Text>An action rule&rsquo;s name must be unique.</Form.Text>
+      </td>
+      <td>
+        <Form.Control as="textarea" rows={4} required ref={mustHaveLabelsRef} />
+      </td>
+      <td>
+        <Form.Control as="textarea" rows={4} ref={mustNotHaveLabelsRef} />
+      </td>
+      <td>
+        <Form.Control as="select" required ref={actionRef}>
+          <option value="0" key="0" selected>
+            Select an action...
+          </option>
+          {actionOptions}
+        </Form.Control>
+      </td>
+    </>
+  );
+}
 
 export default function ActionRuleSettingsTab() {
+  const [adding, setAdding] = useState(false);
+  const [nameRef] = useState(React.createRef());
```

Comment:
Not following this. Is it a common pattern now to have refs as 'state'?

I have seen `useRef()` in the past, and simply `const x = React.createRef()`, but this is new.

Ofcourse, my knowledge of react may have already been outdated. 😓

Reply:
no longer applicable. reworked to not use refs.